### DISCUSSION TITLE
Metadata with HLS and %fdkaac?

### DISCUSSION THREAD

**Discussion by JonathanThorpe** [savonet/liquidsoap#3602] [General]
Hi All,

HLS on LS 2.2.2 works really well, but I noticed that when using the %fdkaac encoder, I don't seem to get metadata on those streams?

This is what I have:
```
aac_lofi = %fdkaac(channels=2,
                    samplerate=48000,
                    bandwidth="auto",
                    bitrate=32,
                    afterburner=true,
                    aot="mpeg4_he_aac_v2",
                    transmux="adts",
                    sbr_mode=true)

  aac_midfi = %fdkaac(channels=2,
                    samplerate=48000,
                    bandwidth="auto",
                    bitrate=64,
                    afterburner=true,
                    aot="mpeg4_he_aac",
                    transmux="adts",
                    sbr_mode=true)

  aac_hifi = %ffmpeg(format="adts",
                    %audio(
                      codec="aac",
                      channels=2,
                      ar=48000,
                      b="128k"
                    )).{id3=true}

  streams = [("aac_lofi", aac_lofi),
            ("aac_midfi", aac_midfi),
            ("aac_hifi", aac_hifi)]

   hls_segments_directory = "/var/hls-tmp/#{station_hls}/live"
  persist_dir = "/srv/liquidsoap-stereotool/hls-persist/#{station_hls}"
  temp_dir = "/var/hls-tmp/#{station_hls}/temp"

  output.file.hls(playlist="#{station_hls}.m3u8",
                  hls_segments_directory,
                  temp_dir=temp_dir,
                  persist_at="#{persist_dir}/persist.config",
                  segment_duration=4.0,
                  segments=10,
                  segments_overhead=20,
                  segment_name=segment_name,
                  streams,
                  hls_source)

```

The aac_hifi stream encoded with ffmpeg shows metadata as expected, but the %fdkaac does not. Is there something I've missed?

I also tried using the ffmpeg he-aac encoder as below:
```
  aac_lofi = %ffmpeg(format="adts",
                    %audio(
                      codec="libfdk_aac",
                      samplerate=48000,
                      b="32k",
                      afterburner=1,
                      profile="aac_he_v2"
                    )).{id3=true}

  aac_midfi = %ffmpeg(format="adts",
                    %audio(
                      codec="libfdk_aac",
                      samplerate=48000,
                      b="64k",
                      afterburner=1,
                      #profile="he_aac"
                    )).{id3=true}
```

This works with metadata, however my m3u8 file seems to show that all of them are encoded as `mp4a.40.2` which isn't right?

**Comment by gAlleb** [comment]
https://www.liquidsoap.info/doc-dev/hls_output.html
```
HLS outputs supports metadata in two ways:

    Through a timed_id3 metadata logical stream with the mpegts format.
    Through regular ID3 frames, as requested by the [HLS specifications](https://datatracker.ietf.org/doc/html/rfc8216#section-3.4) for adts, mp3, ac3 and eac3 formats.
    There is currently no support for in-stream metadata for the mp4 format
```
Hi @JonathanThorpe!

1. So why in you first example you are using `%fdkaac`?
```
  aac_midfi = %fdkaac(channels=2,
                    samplerate=48000,
                    bandwidth="auto",
                    bitrate=64,
                    afterburner=true,
                    aot="mpeg4_he_aac",
                    transmux="adts",
                    sbr_mode=true)
```
Why use external fdkaac if you have a great opportunity of using it inside ffmpeg (the problem with metadata may be hidden here)? If you you have ffmpeg compiled with `fdkaac` (which I see from your second example that you have already did) then just use it inside ffmpeg (codec="libfdk_aac") and thus metadata may appear, don't forget `id3_version = 4` like this:
```
 aac_lofi = %ffmpeg(format="adts",
                    %audio(
                      codec="libfdk_aac",
                      channels=2,
                      samplerate=48000,
                      b="32",
                      afterburner=1,
                      profile="aac_he_v2"
                    )).{
                      id3_version = 4
                        }
 aac_midfi = %ffmpeg(format="adts",
                    %audio(
                      codec="libfdk_aac",
                      channels=2,
                      samplerate=48000,
                      b="64",
                      afterburner=1,
                      profile="aac_he_v2"
                    )).{
                      id3_version = 4
                        }
```
And maybe change format to `mpegts` for best compatibility + it supports `timed_id3` metadata vs regular `ID3 frames` for `adts`.

2. Regarding `mp4a.40.2`. I might be mistaken but may be its because you have also `#profile="he_aac"` in aac_midfi (no :) ).
Also there is an idea that its just a `bug` :) 

You could try and `ffprobe` against the generated MPEG-TS file (`ffprobe filename.ts`) — it might give you the correct result (`aac (HE-AACv2)`) so there are no worries. Something like this I hope:

```
Input #0, mpegts, from 'fm_midfi_3_1702555774_973.ts':
  Duration: 00:00:03.04, start: 2939.785578, bitrate: 223 kb/s
  Program 1 
    Metadata:
      service_name    : Service01
      service_provider: FFmpeg
  Stream #0:0[0x100]: Audio: aac (HE-AACv2) ([15][0][0][0] / 0x000F), 48000 Hz, stereo, fltp, 162 kb/s
  Stream #0:1[0x101]: Data: timed_id3 (ID3  / 0x20334449)
```

**Comment by toots** [comment]
@JonathanThorpe although not properly documented, id3 in HLS output are currently only supported when using the `%ffmpeg` encoder. https://github.com/savonet/liquidsoap/pull/3604 should add support to the `%mp3`, `%shine` and `%fdkaac` as well.

Regarding the audio codec attribute, this is a bug in `ocaml-ffmpeg` (and ffmpeg too!), which should be fixed in https://github.com/savonet/ocaml-ffmpeg/commit/2bf0edc879fe7958e024f3ff91e2280409b75924

I'll see about pushing this to proper stable release depending on timelines for each branch/project.

**Reply by toots** [reply]
I forgot: for the codec attribute, you should be able to specify it manually meanwhile.

**Comment by JonathanThorpe** [comment]
Thanks @toots  for the reply - very helpful!

Regarding the timed_id3 format, does this not also require the use of the "X-PROGRAM-DATE-TIME" as well? This doesn't appear to be auto-generated - I expect I need to be able to generate these every time metadata is updated?

Also, regarding specifying the codec attribute manually, I was able to set this successfully with the "codecs" attribute on each stream.

One other thing I've noticed when using mpegts when using the Apple Media Stream Validator is that it complains that the "Measured peak bitrate compared to multivariant playlist declared value exceeds error tolerance". Doing an ffprobe against some of the .ts files (as you've done above), you can see that the bitrate for the file is substantially higher than the bitrate of the stream itself. Is this expected?

With:
```
aac_hifi = %ffmpeg(format="mpegts",
                    %audio(
                      codec="aac",
                      channels=2,
                      ar=48000,
                      b="128k"
                    )).{id3_version = 4}
```

I get something like:
```
Input #0, mpegts, from 'aac_aac_hifi_4_390_1705373332.ts':
  Duration: 00:00:04.05, start: 1550.784000, bitrate: 215 kb/s
  Program 1 
    Metadata:
      service_name    : Service01
      service_provider: FFmpeg
  Stream #0:0[0x100]: Audio: aac (LC) ([15][0][0][0] / 0x000F), 48000 Hz, stereo, fltp, 135 kb/s
  Stream #0:1[0x101]: Data: timed_id3 (ID3  / 0x20334449)
```

Where does the 215 kb/s bitrate come from here?

**Reply by JonathanThorpe** [reply]
Perhaps to answer my own question, and maybe to help anyone else who comes across this: it appears that the mpegts container format is not very efficient and on average adds around ~64k to each segment. ADTS seems to be a lot more efficient in this regard, but I still had to tune the bandwidth parameter when using the "aac" codec for the aac_hifi stream.

**Reply by toots** [reply]
Yeah, `mpegts` standard has some bad overhead. If ADTS works for you it's a better alternative. However, last we checked, the format with the most cross-device compatiblity was `mpegts`. In particular, android has a pretty limited, hardcoded support for HLS. At least that was the case the last time I checked.
